core/src/main/java/hudson/PluginManager.java

```diff
@@ -423,6 +434,160 @@ public void run(Reactor reactor) throws Exception {
+                LOGGER.log(Level.SEVERE, "Failed to extract the bundled plugin "+fileName,e);
+            }
+        }
+        
+        // Copy dependencies. These are not detached plugins, but are required by them.
+        for (URL dependency : dependencies) {
+            if (copiedPlugins.contains(dependency)) {
+                // Ignore. Already copied.
+                continue;
+            }
+            
+            String fileName = new File(dependency.getFile()).getName();
+            try {
+                names.add(fileName);
+                copyBundledPlugin(dependency, fileName);
+                copiedPlugins.add(dependency);
+            } catch (IOException e) {
+                LOGGER.log(Level.SEVERE, "Failed to extract the bundled dependency plugin " + fileName, e);
+            }
+        }
+        
+        return names;
+    }
+
+    protected static void addDependencies(URL hpiResUrl, String fromPath, Set<URL> dependencySet) throws URISyntaxException, MalformedURLException {
+        if (dependencySet.contains(hpiResUrl)) {
+            return;
+        }
+        
+        Manifest manifest = parsePluginManifest(hpiResUrl);
+        String dependencySpec = manifest.getMainAttributes().getValue("Plugin-Dependencies");
+        if (dependencySpec != null) {
+            String[] dependencyTokens = dependencySpec.split(",");
+            ServletContext context = Jenkins.getInstance().servletContext;
+
+            for (String dependencyToken : dependencyTokens) {
+                if (dependencyToken.endsWith(";resolution:=optional")) {
+                    // ignore optional dependencies
+                    continue;
+                }
+
+                String artifactId = dependencyToken.split(":")[0];
+                URL dependencyURL = context.getResource(fromPath + "/" + artifactId + ".hpi");
+
+                if (dependencyURL == null) {
+                    // Maybe bundling has changed .jpi files
+                    dependencyURL = context.getResource(fromPath + "/" + artifactId + ".jpi");
+                }
+
+                if (dependencyURL != null) {
+                    dependencySet.add(dependencyURL);
+                    // And transitive deps...
+                    addDependencies(dependencyURL, fromPath, dependencySet);
+                }
+            }
+        }
+    }
+
+    /**
+     * Load detached plugins and their dependencies.
+     * <p>
+     * Only loads plugins that:
+     * <ul>
+     *     <li>Have been detached since the last running version.</li>
+     *     <li>Are already installed and need to be upgraded. This can be the case if this Jenkins install has been running since before plugins were "unbundled".</li>
+     *     <li>Are dependencies of one of the above e.g. script-security is not one of the detached plugins but it must be loaded if matrix-project is loaded.</li>
+     * </ul>
+     */
+    protected void loadDetachedPlugins() {
```

Comment:
@stephenc I thought it was doing something different and it didn't seem to handle plugin dependencies. There are no docs with it so reading the code is the only way of making sense of it - maybe I didn't read it closely enough.